driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -62,9 +59,8 @@
     @VisibleForTesting
     final Set<Connection> trash = new CopyOnWriteArraySet<Connection>();
 
-    private volatile int waiter = 0;
-    private final Lock waitLock = new ReentrantLock(true);
-    private final Condition hasAvailableConnection = waitLock.newCondition();
+    private final Queue<SettableFuture<Connection>> pendingBorrows = new ConcurrentLinkedQueue<SettableFuture<Connection>>();
+    private final AtomicInteger pendingBorrowCount = new AtomicInteger();
```

Comment:
It also reminded me some bad memories from `EventDebouncer` :) but indeed I think the counter is required because `size()` is O(N). I think the counter and the queue could be out of sync momentarily but I did not find any situation where this would lead to a bad behavior of the pool.